File: KotlinProject/composeApp/src/jvmMain/kotlin/org/example/project/core/utils/ErrorLogger.kt
package org.example.project.core.utils

/**
 * Standardized error logging utility for the WordBridge application.
 *
 * All errors are logged in the format: [FileName]{ErrorMessage}
 *
 * This ensures:
 * - Clear indication of the source file where the error originated
 * - Consistent logging across all modules and layers
 * - No exposure of sensitive information (sanitized messages)
 *
 * Usage:
 * ```
 * ErrorLogger.log("MyClass.kt", "Failed to load data")
 * ErrorLogger.logException("MyClass.kt", exception)
 * ErrorLogger.logWarning("MyClass.kt", "Connection timeout")
 * ```
 */
object ErrorLogger {
    // Sensitive patterns to sanitize from error messages
    private val sensitivePatterns =
        listOf(
            Regex("""password[=:]\s*\S+""", RegexOption.IGNORE_CASE),
            Regex("""token[=:]\s*\S+""", RegexOption.IGNORE_CASE),
            Regex("""api[_-]?key[=:]\s*\S+""", RegexOption.IGNORE_CASE),
            Regex("""secret[=:]\s*\S+""", RegexOption.IGNORE_CASE),
            Regex("""authorization[=:]\s*\S+""", RegexOption.IGNORE_CASE),
            Regex("""bearer\s+\S+""", RegexOption.IGNORE_CASE),
            Regex("""\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b"""), // Email pattern
        )

    /**
     * Log an error message with standardized format.
     *
     * @param fileName The source file name where the error occurred (e.g., "AuthViewModel.kt")
     * @param message The error message to log
     */
    fun log(
        fileName: String,
        message: String,
    ) {
        val sanitizedMessage = sanitizeMessage(message)
        println("[$fileName]{$sanitizedMessage}")
    }

    /**
     * Log an exception with standardized format.
     *
     * @param fileName The source file name where the error occurred
     * @param exception The exception that was caught
     * @param additionalContext Optional additional context about the error
     */
    fun logException(
        fileName: String,
        exception: Throwable,
        additionalContext: String? = null,
    ) {
        val baseMessage = exception.message ?: "Unknown error"
        val sanitizedMessage = sanitizeMessage(baseMessage)

        val fullMessage =
            if (additionalContext != null) {
                "$additionalContext: $sanitizedMessage"
            } else {
                sanitizedMessage
            }

        println("[$fileName]{$fullMessage}")
    }

    /**
     * Log a warning message with standardized format.
     *
     * @param fileName The source file name where the warning occurred
     * @param message The warning message to log
     */
    fun logWarning(
        fileName: String,
        message: String,
    ) {
        val sanitizedMessage = sanitizeMessage(message)
        println("[$fileName][WARNING]{$sanitizedMessage}")
    }

    /**
     * Log an info message with standardized format.
     *
     * @param fileName The source file name
     * @param message The info message to log
     */
    fun logInfo(
        fileName: String,
        message: String,
    ) {
        val sanitizedMessage = sanitizeMessage(message)
        println("[$fileName][INFO]{$sanitizedMessage}")
    }

    /**
     * Log a debug message with standardized format.
     * Only logs in debug mode.
     *
     * @param fileName The source file name
     * @param message The debug message to log
     */
    fun logDebug(
        fileName: String,
        message: String,
    ) {
        // Only log debug messages if debug mode is enabled
        if (isDebugMode()) {
            val sanitizedMessage = sanitizeMessage(message)
            println("[$fileName][DEBUG]{$sanitizedMessage}")
        }
    }

    /**
     * Sanitize message to remove any potentially sensitive information.
     *
     * @param message The original message
     * @return Sanitized message with sensitive data replaced
     */
    private fun sanitizeMessage(message: String): String {
        var sanitized = message

        for (pattern in sensitivePatterns) {
            sanitized = pattern.replace(sanitized, "[REDACTED]")
        }

        return sanitized
    }

    /**
     * Check if the application is running in debug mode.
     */
    private fun isDebugMode(): Boolean {
        // Check system property or environment variable
        return System.getProperty("wordbridge.debug")?.toBoolean()
            ?: System.getenv("WORDBRIDGE_DEBUG")?.toBoolean()
            ?: true // Default to true for development
    }
}

/**
 * Extension function for easy error logging from any class.
 * Usage: "MyClass.kt".logError("Something went wrong")
 */
fun String.logError(message: String) {
    ErrorLogger.log(this, message)
}

/**
 * Extension function for easy exception logging from any class.
 * Usage: "MyClass.kt".logException(exception)
 */
fun String.logException(
    exception: Throwable,
    context: String? = null,
) {
    ErrorLogger.logException(this, exception, context)
}

/**
 * Extension function for easy warning logging from any class.
 * Usage: "MyClass.kt".logWarning("Connection slow")
 */
fun String.logWarning(message: String) {
    ErrorLogger.logWarning(this, message)
}

File: backend/error_logger.py
"""
Standardized error logging utility for the WordBridge backend.

All errors are logged in the format: [FileName]{ErrorMessage}

This ensures:
- Clear indication of the source file where the error originated
- Consistent logging across all modules and layers
- No exposure of sensitive information (sanitized messages)

Usage:
    from error_logger import ErrorLogger
    
    Logger = ErrorLogger("my_file.py")
    logger.error("Something went wrong")
    logger.warning("Connection slows")
    logger.info("Processing started")
"""

import re
import logging
from typing import Optional

# Sensitive patterns to sanitize from error messages
SENSITIVE_PATTERNS = [
    re.compile(r'password[=:]\s*\S+', re.IGNORECASE),
    re.compile(r'token[=:]\s*\S+', re.IGNORECASE),
    re.compile(r'api[_-]?key[=:]\s*\S+', re.IGNORECASE),
    re.compile(r'secret[=:]\s*\S+', re.IGNORECASE),
    re.compile(r'authorization[=:]\s*\S+', re.IGNORECASE),
    re.compile(r'bearer\s+\S+', re.IGNORECASE),
    re.compile(r'\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b'),  # Email pattern
]


def sanitize_message(message: str) -> str:
    """Remove sensitive information from error messages."""
    sanitized = message
    for pattern in SENSITIVE_PATTERNS:
        sanitized = pattern.sub('[REDACTED]', sanitized)
    return sanitized


class ErrorLogger:
    """Standardized error logger for the WordBridge backend."""
    
    def __init__(self, file_name: str):
        """
        Initialize the error logger.
        
        Args:
            file_name: The source file name (e.g., "main.py")
        """
        self.file_name = file_name
        self._logger = logging.getLogger(file_name)
    
    def error(self, message: str, exc: Optional[Exception] = None):
        """
        Log an error message with standardized format.
        
        Args:
            message: The error message to log
            exc: Optional exception that caused the error
        """
        sanitized = sanitize_message(message)
        formatted = f"[{self.file_name}]{{{sanitized}}}"
        
        if exc:
            self._logger.error(formatted, exc_info=exc)
        else:
            self._logger.error(formatted)
    
    def warning(self, message: str):
        """
        Log a warning message with standardized format.
        
        Args:
            message: The warning message to log
        """
        sanitized = sanitize_message(message)
        formatted = f"[{self.file_name}][WARNING]{{{sanitized}}}"
        self._logger.warning(formatted)
    
    def info(self, message: str):
        """
        Log an info message with standardized format.
        
        Args:
            message: The info message to log
        """
        sanitized = sanitize_message(message)
        formatted = f"[{self.file_name}][INFO]{{{sanitized}}}"
        self._logger.info(formatted)
    
    def debug(self, message: str):
        """
        Log a debug message with a standardized format.
        
        Args:
            message: The debug message to log
        """
        sanitized = sanitize_message(message)
        formatted = f"[{self.file_name}][DEBUG]{{{sanitized}}}"
        self._logger.debug(formatted)


# Convenience function to create a logger for a file
def get_logger(file_name: str) -> ErrorLogger:
    """
    Get a standardized error logger for a file.
    
    Args:
        file_name: The source file name (e.g., "main.py")
        
    Returns:
        An ErrorLogger instance for the file
    """
    return ErrorLogger(file_name)
